modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift

```diff
@@ -34,304 +39,208 @@ extension AddressBookClient: DependencyKey {
+            storeData = false
+        } catch RemoteStorageClient.RemoteStorageError.containerURL {
+            storeData = false
+        } catch {
+            throw error
         }
 
-        return data
-    }
-    
-    private static func decryptData(_ data: Data) throws -> AddressBookContacts {
-        @Dependency(\.walletStorage) var walletStorage
-
-        // TODO: str4d
-//        guard let encryptionKeys = try? walletStorage.exportAddressBookEncryptionKeys() else {
-//            throw AddressBookClient.AddressBookClientError.missingEncryptionKey
-//        }
-
-        // here you have the encrypted data from the cloud, the blob
-        // you also have a key from the keychain
-        
-        var offset = 0
-
-        // Deserialize `version`
-        let versionBytes = data.subdata(in: offset..<(offset + MemoryLayout<Int>.size))
-        offset += MemoryLayout<Int>.size
-
-        guard let version = AddressBookClient.bytesToInt(Array(versionBytes)) else {
-            return .empty
-        }
-        
-        guard version == AddressBookContacts.Constants.version else {
-            return .empty
-        }
-
-        // Deserialize `lastUpdated`
-        guard let lastUpdated = AddressBookClient.deserializeDate(from: data, at: &offset) else {
-            return .empty
-        }
-
-        // Deserialize `contactsCount`
-        let contactsCountBytes = data.subdata(in: offset..<(offset + MemoryLayout<Int>.size))
-        offset += MemoryLayout<Int>.size
-
-        guard let contactsCount = AddressBookClient.bytesToInt(Array(contactsCountBytes)) else {
-            return .empty
-        }
-        
-        var contacts: [Contact] = []
-        for _ in 0..<contactsCount {
-            if let contact = AddressBookClient.deserializeContact(from: data, at: &offset) {
-                contacts.append(contact)
-            }
-        }
-        
-        let abContacts = AddressBookContacts(
-            lastUpdated: lastUpdated,
+        // Merge strategy
+        var syncedContacts = AddressBookContacts(
+            lastUpdated: Date(),
             version: AddressBookContacts.Constants.version,
-            contacts: IdentifiedArrayOf(uniqueElements: contacts)
+            contacts: contacts.contacts
         )
 
-        return abContacts
-    }
-    
-    private static func serializeContact(_ contact: Contact) -> Data {
-        var data = Data()
-
-        // Serialize `lastUpdated`
-        data.append(contentsOf: AddressBookClient.serializeDate(contact.lastUpdated))
-
-        // Serialize `address` (length + UTF-8 bytes)
-        let addressBytes = stringToBytes(contact.id)
-        data.append(contentsOf: intToBytes(addressBytes.count))
-        data.append(contentsOf: addressBytes)
+        remoteContacts.contacts.forEach {
+            var notFound = true
+            var indexToUpdate = -1
+            
+            for i in 0..<syncedContacts.contacts.count {
```

Comment:
This is $O(n \cdot n')$ where $n$ is the number of remote contacts and $n'$ is the number of synced contacts. If the contacts are sorted then it should only take $O(m \cdot \log m)$ where $m = \max(n, n')$. Swift probably has a sorted map type with a merge operation. It might be [this one](https://developer.apple.com/documentation/swift/dictionary/merge(_:uniquingkeyswith:)-m2ub).

Reply:
#1408 follow up